src/mds/MDSMap.cc

```diff
@@ -687,6 +678,11 @@ MDSMap::availability_t MDSMap::is_cluster_available() const
     return AVAILABLE;
   } else {
     // Nothing indicating we were stuck, but nobody active (yet)
-    return TRANSIENT_UNAVAILABLE;
+    //return TRANSIENT_UNAVAILABLE;
+
+    // Because we don't have standbys in the MDSMap any more, we can't
+    // reliably indicate transient vs. stuck, so always say stuck so
+    // that the client doesn't block.
+    return STUCK_UNAVAILABLE;
```

Comment:
This sounds like a hole: we need older clients without this fix to behave correctly as well, which means they need the appropriate data in a form they can understand. I haven't read through many of the relevant changes here but I think we could do something like force the MDSMonitor to tick() whenever a client tries to connect for a down FS, so that if it's capable of coming back the right number of MDSes will be in the map it's reading.
(Or more likely we'd pull out all that MDS reassignment code into its own function, give it some memory so we don't do it repeatedly if we don't have any standby MDSes, etc, but you get the idea.)